File: src/main/drive.ts
import drivelist from "drivelist";

export interface DriveInfo {
    device: string;
    description: string;
    size: number;
    mountpoints: { path: string }[];
    isSystem: boolean;
    isRemovable: boolean;
}

export async function listDrives(): Promise<DriveInfo[]> {
    try {
        const drives = await drivelist.list();
        // 只返回有挂载点的设备，简化前端处理
        return drives
            .filter((d: unknown) => {
                const drive = d as Record<string, unknown>;
                return (
                    drive.mountpoints &&
                    Array.isArray(drive.mountpoints) &&
                    drive.mountpoints.length > 0
                );
            })
            .map((d: unknown) => {
                const drive = d as Record<string, unknown>;
                return {
                    device: drive.device as string,
                    description: drive.description as string,
                    size: drive.size as number,
                    mountpoints: drive.mountpoints as { path: string }[],
                    isSystem: drive.system as boolean,
                    isRemovable: drive.isRemovable as boolean,
                };
            });
    } catch (err) {
        console.error("[drivelist] error:", err);
        console.error("[drivelist] process.env.PATH:", process.env.PATH);
        return [];
    }
}

File: backend/internal/drive/drive.go
package drive

import (
	"os"
	"path/filepath"
	"sort"
)

// Info describes a mount point or volume.
type Info struct {
	Device      string              `json:"device"`
	Description string              `json:"description"`
	Size        int64               `json:"size"`
	Mountpoints []map[string]string `json:"mountpoints"`
	IsSystem    bool                `json:"isSystem"`
	IsRemovable bool                `json:"isRemovable"`
	IsReady     bool                `json:"isReady"`
	Label       string              `json:"label,omitempty"`
}

// List returns home and mounted volumes.
func List() ([]Info, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return nil, err
	}

	drives := []Info{
		{
			Device:      "home",
			Description: "Home",
			Mountpoints: []map[string]string{{"path": home}},
			IsSystem:    true,
			IsRemovable: false,
			IsReady:     true,
			Label:       "Home",
		},
	}

	volumeRoot := "/Volumes"
	entries, err := os.ReadDir(volumeRoot)
	if err != nil {
		return drives, nil
	}

	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		mountPath := filepath.Join(volumeRoot, entry.Name())
		drives = append(drives, Info{
			Device:      entry.Name(),
			Description: entry.Name(),
			Mountpoints: []map[string]string{{"path": mountPath}},
			IsSystem:    false,
			IsRemovable: true,
			IsReady:     true,
			Label:       entry.Name(),
		})
	}

	sort.Slice(drives, func(i, j int) bool {
		return drives[i].Description < drives[j].Description
	})

	return drives, nil
}

// Get returns one drive by device id.
func Get(device string) (Info, error) {
	drives, err := List()
	if err != nil {
		return Info{}, err
	}

	for _, item := range drives {
		if item.Device == device {
			return item, nil
		}
	}

	return Info{}, os.ErrNotExist
}
